NPE when introspecting JWT access token in tenant qualified URLs disabled environment when build_issuer_with_hostname enable and hostname is changed

### Description

When the tenant qualified URLs are disabled, the issuer value for the JWT access tokens will be taken from the entity ID stored in the database.

If the hostname is changed after the first startup, this value will not be changed and the JWT access tokens will still have the entity ID value stored in the database as the issuer.

The following config can be use in order to build the issuer value from the hostname of the deploymnet.toml file.

```
[oauth]
build_issuer_with_hostname = true
```

However, the tokens retrieved with the above config will fail in the introspection flow throwing the following NPE.

```
[2026-06-03 02:22:24,507] [e5f85362-e098-4788-b9e9-cfffc48f026e] ERROR {org.wso2.carbon.identity.oauth2.OAuth2TokenValidationService} - Error occurred while building the introspection responseorg.wso2.carbon.identity.oauth2.IdentityOAuth2Exception: Error occurred while validating token.
        at org.wso2.carbon.identity.oauth2.validators.TokenValidationHandler.handleTokenValidationException(TokenValidationHandler.java:434)
        at org.wso2.carbon.identity.oauth2.validators.TokenValidationHandler.buildIntrospectionResponse(TokenValidationHandler.java:355)
        at org.wso2.carbon.identity.oauth2.OAuth2TokenValidationService.buildIntrospectionResponse(OAuth2TokenValidationService.java:135)
        at org.wso2.carbon.identity.oauth.endpoint.introspection.OAuth2IntrospectionEndpoint.introspect(OAuth2IntrospectionEndpoint.java:128)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
...
Caused by: java.lang.NullPointerException: Cannot invoke "String.equals(Object)" because "resourceResidentOrgId" is null
        at org.wso2.carbon.identity.oauth2.util.JWTUtils.getIDPForIssuer(JWTUtils.java:219)
        at org.wso2.carbon.identity.oauth2.util.JWTUtils.getResidentIDPForIssuer(JWTUtils.java:252)
        at org.wso2.carbon.identity.oauth2.validators.OAuth2JWTTokenValidator.validateAccessToken(OAuth2JWTTokenValidator.java:103)
        at org.wso2.carbon.identity.oauth2.validators.TokenValidationHandler.buildIntrospectionResponse(TokenValidationHandler.java:311)
        ... 70 more
```

The user will receive the access token not active reponse while the backend logs have the NPE. This has to be handled that, token retrieved with `build_issuer_with_hostname` must be able to introspect without any issue.

### Steps to Reproduce

1. Get a fresh IS pack
2. Disable tenant qualified URLs
```
[tenant_context]
enable_tenant_qualified_urls = false
enable_tenanted_sessions = false
```
3. Add the necessary database constraint for the non tenant qualified URLs
4. Add a hostname to the toml
```
[server]
hostname = "wso2is.dev"
```
5. Start the IS
6. Create an standard OIDC application with JWT access token
7. Get an access token with client credentials grant
8. Issuer of the token will be `https://wso2is.dev:9443/oauth2/token`
9. Stop the IS and change the hostname in toml
```
[server]
hostname = "wso2is.com"
```
10. Start the IS and get an access token with the same application
11. Issuer will not change (`https://wso2is.dev:9443/oauth2/token`)
12. Stop the server and add the following config to the toml
```
[oauth]
build_issuer_with_hostname = true
```
13. Start the IS and get an access token from the same application
14. Issuer will be changed to the new hostname (https://wso2is.com:9443/oauth2/token)
15. Introspect the new access token
16. Response will say the access token is not active.
17. Backend logs will show the NPE.

### Please select the area issue is related to

API Access Mgt & Authorization

### Version

7.2.0

### Environment Details (with versions)

_No response_

### Developer Checklist

- [ ] [Behavioural Change] Does this change introduce a behavioral change to the product?
- [ ]  ↳ Approved by team lead
- [ ]  ↳ Label `impact/behavioral-change` added
- [ ] [Migration Impact] Does this change have a migration impact?
- [ ]  ↳ Migration label added (e.g., `7.2.0-migration`)
- [ ]  ↳ Migration issues created and linked
- [ ] [New Configuration] Does this change introduce a new configuration?
- [ ]  ↳ Label `config` added
- [ ]  ↳ Configuration is properly documented

--- a/components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/JWTUtils.java
+++ b/components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/JWTUtils.java
@@ -205,6 +205,13 @@ public class JWTUtils {
                     jwtIssuer.equals(OAuth2Util.OAuthURL.getOAuth2MTLSTokenEPUrl())) {
                 return residentIdentityProvider;
             }
+            // When the issuer is built with hostname (and tenant qualified URLs disabled), compare against the
+            // recomputed hostname based issuer instead of the stale entity ID stored in the database.
+            if (Boolean.parseBoolean(IdentityUtil.getProperty(OAuthConstants.OAUTH_BUILD_ISSUER_WITH_HOSTNAME)) &&
+                    !IdentityTenantUtil.isTenantQualifiedUrlsEnabled() &&
+                    jwtIssuer.equals(OAuth2Util.getIssuerLocation(tenantDomain))) {
+                return residentIdentityProvider;
+            }
             // Check for organization management enablement.
             if (!OAuth2ServiceComponentHolder.getInstance().isOrganizationManagementEnabled()) {
                 throw new IdentityOAuth2Exception("No registered IDP found for the token with issuer name : "
@@ -221,6 +228,11 @@ public class JWTUtils {
             }
             int depthOfRootOrg = getSubOrgStartLevel() - 1;
             String resourceResidentOrgId = PrivilegedCarbonContext.getThreadLocalCarbonContext().getOrganizationId();
+            // Bail out early if the organization IDs could not be resolved to avoid a NullPointerException.
+            if (jwtIssuerOrgId == null || resourceResidentOrgId == null) {
+                throw new IdentityOAuth2Exception("No registered IDP found for the token with issuer name : "
+                        + jwtIssuer);
+            }
             if (!jwtIssuerOrgId.equals(switchedOrgOrgAncestors.get(depthOfRootOrg)) ||
                     !resourceResidentOrgId.equals(switchedOrgId)) {
                 throw new IdentityOAuth2ClientException("No registered IDP found for the token with issuer name : "
@@ -276,6 +288,13 @@ public class JWTUtils {
                     jwtIssuer.equals(OAuth2Util.OAuthURL.getOAuth2MTLSTokenEPUrl())) {
                 return residentIdentityProvider;
             }
+            // When the issuer is built with hostname (and tenant qualified URLs disabled), compare against the
+            // recomputed hostname based issuer instead of the stale entity ID stored in the database.
+            if (Boolean.parseBoolean(IdentityUtil.getProperty(OAuthConstants.OAUTH_BUILD_ISSUER_WITH_HOSTNAME)) &&
+                    !IdentityTenantUtil.isTenantQualifiedUrlsEnabled() &&
+                    jwtIssuer.equals(OAuth2Util.getIssuerLocation(tenantDomain))) {
+                return residentIdentityProvider;
+            }
             // Check for organization management enablement.
             if (!OAuth2ServiceComponentHolder.getInstance().isOrganizationManagementEnabled()) {
                 throw new IdentityOAuth2Exception("No registered IDP found for the token with issuer name : "
@@ -321,6 +340,11 @@ public class JWTUtils {
                         "Ancestor list size is insufficient for the organization ID: " + switchedOrgId);
             }
             String resourceResidentOrgId = PrivilegedCarbonContext.getThreadLocalCarbonContext().getOrganizationId();
+            // Bail out early if the organization IDs could not be resolved to avoid a NullPointerException.
+            if (jwtIssuerOrgId == null || resourceResidentOrgId == null) {
+                throw new IdentityOAuth2Exception("No registered IDP found for the token with issuer name : "
+                        + jwtIssuer);
+            }
             if (!jwtIssuerOrgId.equals(switchedOrgOrgAncestors.get(depthOfRootOrg)) ||
                     !resourceResidentOrgId.equals(switchedOrgId)) {
                 throw new IdentityOAuth2ClientException("No registered IDP found for the token with issuer name : "

--- a/components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/JWTUtilsTest.java
+++ b/components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/util/JWTUtilsTest.java
@@ -26,18 +26,30 @@ import org.testng.annotations.Listeners;
 import org.testng.annotations.Test;
 import org.wso2.carbon.base.CarbonBaseConstants;
 import org.wso2.carbon.context.PrivilegedCarbonContext;
+import org.wso2.carbon.identity.application.common.model.FederatedAuthenticatorConfig;
 import org.wso2.carbon.identity.application.common.model.IdentityProvider;
+import org.wso2.carbon.identity.application.common.model.Property;
 import org.wso2.carbon.identity.application.common.util.IdentityApplicationConstants;
 import org.wso2.carbon.identity.core.IdentityKeyStoreResolver;
 import org.wso2.carbon.identity.core.util.IdentityKeyStoreResolverConstants;
+import org.wso2.carbon.identity.core.util.IdentityTenantUtil;
+import org.wso2.carbon.identity.core.util.IdentityUtil;
 import org.wso2.carbon.identity.oauth.config.OAuthServerConfiguration;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.internal.OAuth2ServiceComponentHolder;
+import org.wso2.carbon.identity.organization.management.service.OrganizationManager;
+import org.wso2.carbon.identity.organization.management.service.util.OrganizationManagementConfigUtil;
+import org.wso2.carbon.identity.organization.management.service.util.OrganizationManagementUtil;
+import org.wso2.carbon.idp.mgt.IdentityProviderManager;
 
 import java.io.FileInputStream;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.security.KeyStore;
 import java.security.cert.X509Certificate;
+import java.util.Arrays;
 
+import static org.mockito.ArgumentMatchers.anyString;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.mockStatic;
 import static org.mockito.Mockito.when;
@@ -78,6 +90,9 @@ public class JWTUtilsTest {
             "3pDI4B+vQipcfTjJqgPAWvUz903z61lRuAlJFPPD69l+IQ15z1Mfc7rgl0wmZLBU\n" +
             "HjVlGsFGQX6e10Rx/msN+NWxKJGf7Z6vxS8Qoc4nddUBnndCOvCVvgI9BThNv0cG\n" +
             "e3hB2nvCQvUJ/wfuj6i1PNfoM81nA2qEQfjY/QWuF4Ex/RYWBfASNU35TBRVc26R";
+
+    private static final String MUTUAL_TLS_ALIASES_ENABLED = "OAuth.MutualTLSAliases.Enabled";
+    private static final String OAUTH_BUILD_ISSUER_WITH_HOSTNAME = "OAuth.BuildIssuerWithHostname";
 
     private PrivilegedCarbonContext privilegedCarbonContext;
 
@@ -127,6 +142,126 @@ public class JWTUtilsTest {
         certificate = JWTUtils.resolveSignerCertificate(identityProvider);
         assertEquals(certificate.getIssuerDN().getName(),
                 "CN=Intermediate CA, OU=IAM, O=WSO2, L=Colombo, ST=Western, C=LK");
+    }
+
+    @Test(description = "getIDPForIssuer returns the resident IDP when the issuer is built with hostname")
+    public void testGetIDPForIssuerWithHostnameBuiltIssuer() throws Exception {
+
+        String hostnameIssuer = "https://localhost:9443/oauth2/token";
+        IdentityProvider residentIdP = buildResidentIdP("https://stale-entity-id");
+        try (MockedStatic<IdentityProviderManager> identityProviderManager = mockStatic(IdentityProviderManager.class);
+             MockedStatic<IdentityUtil> identityUtil = mockStatic(IdentityUtil.class);
+             MockedStatic<IdentityTenantUtil> identityTenantUtil = mockStatic(IdentityTenantUtil.class);
+             MockedStatic<OAuth2Util> oAuth2Util = mockStatic(OAuth2Util.class)) {
+            mockResidentIdP(identityProviderManager, residentIdP);
+            identityUtil.when(() -> IdentityUtil.getProperty(MUTUAL_TLS_ALIASES_ENABLED)).thenReturn("false");
+            identityUtil.when(() -> IdentityUtil.getProperty(OAUTH_BUILD_ISSUER_WITH_HOSTNAME)).thenReturn("true");
+            identityTenantUtil.when(IdentityTenantUtil::isTenantQualifiedUrlsEnabled).thenReturn(false);
+            oAuth2Util.when(() -> OAuth2Util.getIssuerLocation(SUPER_TENANT_DOMAIN_NAME)).thenReturn(hostnameIssuer);
+
+            IdentityProvider result =
+                    JWTUtils.getIDPForIssuer(hostnameIssuer, SUPER_TENANT_DOMAIN_NAME, "switched-org-id");
+            assertEquals(result, residentIdP);
+        }
+    }
+
+    @Test(description = "getResidentIDPIssuer returns the resident IDP when the issuer is built with hostname")
+    public void testGetResidentIDPIssuerWithHostnameBuiltIssuer() throws Exception {
+
+        String hostnameIssuer = "https://localhost:9443/oauth2/token";
+        IdentityProvider residentIdP = buildResidentIdP("https://stale-entity-id");
+        try (MockedStatic<IdentityProviderManager> identityProviderManager = mockStatic(IdentityProviderManager.class);
+             MockedStatic<IdentityUtil> identityUtil = mockStatic(IdentityUtil.class);
+             MockedStatic<IdentityTenantUtil> identityTenantUtil = mockStatic(IdentityTenantUtil.class);
+             MockedStatic<OAuth2Util> oAuth2Util = mockStatic(OAuth2Util.class)) {
+            mockResidentIdP(identityProviderManager, residentIdP);
+            identityUtil.when(() -> IdentityUtil.getProperty(MUTUAL_TLS_ALIASES_ENABLED)).thenReturn("false");
+            identityUtil.when(() -> IdentityUtil.getProperty(OAUTH_BUILD_ISSUER_WITH_HOSTNAME)).thenReturn("true");
+            identityTenantUtil.when(IdentityTenantUtil::isTenantQualifiedUrlsEnabled).thenReturn(false);
+            oAuth2Util.when(() -> OAuth2Util.getIssuerLocation(SUPER_TENANT_DOMAIN_NAME)).thenReturn(hostnameIssuer);
+
+            IdentityProvider result = JWTUtils.getResidentIDPIssuer(hostnameIssuer, "client-id",
+                    SUPER_TENANT_DOMAIN_NAME, "switched-org-id");
+            assertEquals(result, residentIdP);
+        }
+    }
+
+    @Test(description = "getIDPForIssuer throws when the organization IDs cannot be resolved",
+            expectedExceptions = IdentityOAuth2Exception.class)
+    public void testGetIDPForIssuerWithUnresolvableOrgIds() throws Exception {
+
+        IdentityProvider residentIdP = buildResidentIdP("https://stale-entity-id");
+        try (MockedStatic<IdentityProviderManager> identityProviderManager = mockStatic(IdentityProviderManager.class);
+             MockedStatic<IdentityUtil> identityUtil = mockStatic(IdentityUtil.class);
+             MockedStatic<OAuth2ServiceComponentHolder> holder = mockStatic(OAuth2ServiceComponentHolder.class);
+             MockedStatic<OrganizationManagementConfigUtil> configUtil =
+                     mockStatic(OrganizationManagementConfigUtil.class)) {
+            mockResidentIdP(identityProviderManager, residentIdP);
+            identityUtil.when(() -> IdentityUtil.getProperty(MUTUAL_TLS_ALIASES_ENABLED)).thenReturn("false");
+            identityUtil.when(() -> IdentityUtil.getProperty(OAUTH_BUILD_ISSUER_WITH_HOSTNAME)).thenReturn("false");
+            mockOrganizationManager(holder, configUtil);
+
+            JWTUtils.getIDPForIssuer("https://some-issuer", SUPER_TENANT_DOMAIN_NAME, "switched-org-id");
+        }
+    }
+
+    @Test(description = "getResidentIDPIssuer throws when the organization IDs cannot be resolved",
+            expectedExceptions = IdentityOAuth2Exception.class)
+    public void testGetResidentIDPIssuerWithUnresolvableOrgIds() throws Exception {
+
+        IdentityProvider residentIdP = buildResidentIdP("https://stale-entity-id");
+        try (MockedStatic<IdentityProviderManager> identityProviderManager = mockStatic(IdentityProviderManager.class);
+             MockedStatic<IdentityUtil> identityUtil = mockStatic(IdentityUtil.class);
+             MockedStatic<OAuth2ServiceComponentHolder> holder = mockStatic(OAuth2ServiceComponentHolder.class);
+             MockedStatic<OrganizationManagementUtil> organizationManagementUtil =
+                     mockStatic(OrganizationManagementUtil.class);
+             MockedStatic<OrganizationManagementConfigUtil> configUtil =
+                     mockStatic(OrganizationManagementConfigUtil.class)) {
+            mockResidentIdP(identityProviderManager, residentIdP);
+            identityUtil.when(() -> IdentityUtil.getProperty(MUTUAL_TLS_ALIASES_ENABLED)).thenReturn("false");
+            identityUtil.when(() -> IdentityUtil.getProperty(OAUTH_BUILD_ISSUER_WITH_HOSTNAME)).thenReturn("false");
+            organizationManagementUtil.when(() -> OrganizationManagementUtil.isOrganization(SUPER_TENANT_DOMAIN_NAME))
+                    .thenReturn(false);
+            mockOrganizationManager(holder, configUtil);
+
+            JWTUtils.getResidentIDPIssuer("https://some-issuer", "client-id", SUPER_TENANT_DOMAIN_NAME,
+                    "switched-org-id");
+        }
+    }
+
+    private void mockResidentIdP(MockedStatic<IdentityProviderManager> identityProviderManager,
+                                 IdentityProvider residentIdP) throws Exception {
+
+        IdentityProviderManager identityProviderManagerInstance = mock(IdentityProviderManager.class);
+        identityProviderManager.when(IdentityProviderManager::getInstance).thenReturn(identityProviderManagerInstance);
+        when(identityProviderManagerInstance.getResidentIdP(SUPER_TENANT_DOMAIN_NAME)).thenReturn(residentIdP);
+    }
+
+    private void mockOrganizationManager(MockedStatic<OAuth2ServiceComponentHolder> holder,
+                                         MockedStatic<OrganizationManagementConfigUtil> configUtil) throws Exception {
+
+        OAuth2ServiceComponentHolder holderInstance = mock(OAuth2ServiceComponentHolder.class);
+        holder.when(OAuth2ServiceComponentHolder::getInstance).thenReturn(holderInstance);
+        when(holderInstance.isOrganizationManagementEnabled()).thenReturn(true);
+        OrganizationManager organizationManager = mock(OrganizationManager.class);
+        when(holderInstance.getOrganizationManager()).thenReturn(organizationManager);
+        when(organizationManager.resolveOrganizationId(SUPER_TENANT_DOMAIN_NAME)).thenReturn("jwt-issuer-org-id");
+        when(organizationManager.getAncestorOrganizationIds(anyString()))
+                .thenReturn(Arrays.asList("root-org-id", "child-org-id"));
+        configUtil.when(() -> OrganizationManagementConfigUtil.getProperty(anyString())).thenReturn("2");
+    }
+
+    private IdentityProvider buildResidentIdP(String entityId) {
+
+        Property entityIdProperty = new Property();
+        entityIdProperty.setName("IdPEntityId");
+        entityIdProperty.setValue(entityId);
+        FederatedAuthenticatorConfig oidcConfig = new FederatedAuthenticatorConfig();
+        oidcConfig.setName(IdentityApplicationConstants.Authenticator.OIDC.NAME);
+        oidcConfig.setProperties(new Property[]{entityIdProperty});
+        IdentityProvider residentIdP = new IdentityProvider();
+        residentIdP.setFederatedAuthenticatorConfigs(new FederatedAuthenticatorConfig[]{oidcConfig});
+        return residentIdP;
     }
 
     private void mockKeystores() throws Exception {

